storage/s3.go

```diff
@@ -382,6 +386,175 @@ func (s *S3) Get(
+	defer resp.EventStream.Close()
+
+	reader, writer := io.Pipe()
+	defer reader.Close()
+	go func() {
+		defer writer.Close()
+		// TODO: wire ctx up to eventStream.StreamCloser.Close
+		for event := range resp.EventStream.Reader.Events() {
+			switch e := event.(type) {
+			case *s3.RecordsEvent:
+				// TODO: how should we handle errors here?!
+				writer.Write(e.Payload)
+			// How the heck are these different?
+			case *s3.StatsEvent:
+			case *s3.ProgressEvent:
+			}
+		}
+	}()
+
+	decoder := json.NewDecoder(reader)
+	// decoder.DontCopyRawMessage()
+	for {
+		var record json.RawMessage
+		err := decoder.Decode(&record)
+		if err != nil {
+			if err != io.EOF {
+				resultch <- &SelectRecord{
+					Err: err,
+				}
+			}
+			return
+		}
+		resultch <- &SelectRecord{
+			Record: record,
+		}
+	}
+	err := resp.EventStream.Err()
+	if err != nil {
+		resultch <- &SelectRecord{
+			Err: err,
+		}
+	}
+}
+
+func (s *S3) selectWorker(ctx context.Context, input s3.SelectObjectContentInput, objch <-chan *Object, resultch chan<- *SelectRecord) {
+	for object := range objch {
+		// TODO: reintroduce
+		if object.Type.IsDir() { // || errorpkg.IsCancelation(object.Err) {
+			continue
+		}
+
+		if err := object.Err; err != nil {
+			resultch <- &SelectRecord{
+				Err: object.Err,
+			}
+			continue
+		}
+
+		if object.StorageClass.IsGlacier() {
+			// TODO: this isn't fatal - how to signal?
+			resultch <- &SelectRecord{
+				Err: fmt.Errorf("object '%v' is on Glacier storage", object),
+			}
+			continue
+		}
+
+		url := object.URL
+		input.Bucket = aws.String(url.Bucket)
+		input.Key = aws.String(url.Path)
+
+		out, err := s.api.SelectObjectContentWithContext(ctx, &input)
+		if err != nil {
+			resultch <- &SelectRecord{
+				Err: err,
+			}
+		} else {
+			s.selectStream(out, resultch)
+		}
+	}
+}
+
+type Query struct {
+	ExpressionType  string
+	Expression      string
+	CompressionType string
+}
+
+func writev(to io.Writer, vector ...[]byte) (err error) {
+	for _, vec := range vector {
+		_, err = to.Write(vec)
+		if err != nil {
+			break
+		}
+	}
+	return
+}
+
+func (s *S3) Select(ctx context.Context, objch <-chan *Object, to io.Writer, query *Query, concurrency int) error {
+	if s.dryRun {
+		return nil
+	}
+
+	baseInput := s3.SelectObjectContentInput{
+		ExpressionType: aws.String(query.ExpressionType),
+		Expression:     aws.String(query.Expression),
+		InputSerialization: &s3.InputSerialization{
+			CompressionType: aws.String(query.CompressionType),
+			JSON: &s3.JSONInput{
+				Type: aws.String("Lines"),
+			},
+		},
+		OutputSerialization: &s3.OutputSerialization{
+			JSON: &s3.JSONOutput{},
+		},
+		// RequestProgress: &s3.RequestProgress{
+		// 	// TODO: conditionally enable? for larger objects? ?
+		// 	Enabled: false,
+		// },
+	}
+
+	workerCtx, cancel := context.WithCancel(ctx)
+
+	resultch := make(chan *SelectRecord, concurrency)
+	var wg sync.WaitGroup
+	wg.Add(concurrency)
```

Comment:
There's a natural split in the design of this command: some `s5cmd select` calls will match very few lines out of a lot of files, and would benefit significantly from more goroutines. However, some calls match many lines across fewer files, and in the interest of avoiding excess context switching it doesn't make sense to have the same huge number of goroutines there. Maybe there's a way to set this up such that there are fewer reader goroutines and many "discovery" goroutines, where the readers actually do the `json` message blob extraction and the discovery goroutines handle sending the `SelectObjectContent` request? I worry that in some cases that'll cause a significant delay between when we make the request and start reading from it, and that that'll cause timeouts on the S3 side.

Reply:
I understand your concern but I don't have a solution other than introducing yet another flag to change *discovery* and *consumer* goroutine count.

Reply:
I don't think we need to solve this now; just commenting that this functionality is relatively important to make this command reasonably support a variety of use-cases (across the fraction space representing data scanned over data retrieved).